### PR TITLE
#1828 Use update method when nested property of target is populated with multiple nested properties of source

### LINKED ISSUE #1828: MapStruct 1.3.0.Final generates invalid implementation when nested property of target is populated with multiple nested properties of source
MapStruct v1.3.0.Final generates invalid mapper implementation when nested object property of target is populated from multiple nested object propertiese of source. MapStruct v1.1.0.Final works fine in this scenario.

```
@Mapper
public interface FirstMapper {

    @Mapping(target = "name", source = "name")
    @Mapping(target = "completeAddress.lineOne", source = "specialAddress.line1")
    @Mapping(target = "completeAddress.lineTwo", source = "specialAddress.line2")
    @Mapping(target = "completeAddress.city", source = "generalAddress.city")
    @Mapping(target = "completeAddress.country", source = "generalAddress.country")
    public Person mapPerson(Employee employee);
}
```
Where structure of target (Person) and person(Employee) is as below
```
class Person { String name; CompleteAddress completeAddress; }

class Employee { String name; GeneralAddress generalAddress,  SpecialAddress specialAddress; }
```

Generated Mapper is as below
```
public class FirstMapperImpl implements FirstMapper {

    @Override
    public Person mapPerson(Employee employee) {
        if ( employee == null ) {
            return null;
        }

        Person person = new Person();

        person.setCompleteAddress( generalAddressToCompleteAddress( employee.getGeneralAddress() ) );
        person.setCompleteAddress( specialAddressToCompleteAddress( employee.getSpecialAddress() ) );
        person.setName( employee.getName() );

        return person;
    }

    protected CompleteAddress generalAddressToCompleteAddress(GeneralAddress generalAddress) {
        if ( generalAddress == null ) {
            return null;
        }

        CompleteAddress completeAddress = new CompleteAddress();

        completeAddress.setCountry( generalAddress.getCountry() );
        completeAddress.setCity( generalAddress.getCity() );

        return completeAddress;
    }

    protected CompleteAddress specialAddressToCompleteAddress(SpecialAddress specialAddress) {
        if ( specialAddress == null ) {
            return null;
        }

        CompleteAddress completeAddress = new CompleteAddress();

        completeAddress.setLineOne( specialAddress.getLine1() );
        completeAddress.setLineTwo( specialAddress.getLine2() );

        return completeAddress;
    }
}
```

Here generated methods `generalAddressToCompleteAddress` and `specialAddressToCompleteAddress` are both returning new instance of `CompleteAddress`. Overridden method `mapPerson` is calling `person.setCompleteAddress` with output of these two generated methods so one instance is thrown away in the process.So, with v1.3.0.Final, mapper is not correctly implemented as per defined mapping.

Please see [this repo](https://github.com/mabilalmirza/mapstruct-multiple-nested-objects) created to demostrate this issue.
- Branch [mapstruct-1.1.0.Final](https://github.com/mabilalmirza/mapstruct-multiple-nested-objects/tree/mapstruct-1.1.0.Final) is passing all unit tests.
- Branch [mapstruct-1.3.0.Final](https://github.com/mabilalmirza/mapstruct-multiple-nested-objects/tree/mapstruct-1.3.0.Final) fails to pass the tests.

### PR DESCRIPTION
**Pull request by filiphr**
Fixes #1828 for the 1.3.1 release

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/model/NestedTargetPropertyMappingHolder.java [hunk 1]
```diff
@@ -160,6 +160,15 @@ public NestedTargetPropertyMappingHolder build() {
                         groupedByTP.singleTargetReferences.get( targetProperty )
                     );
 
+                    // If there are multi source parameters that are mapped to the target reference or
+                    // there are multiple source references to the target reference then we need to force an
+                    // update method. The reason for this is that multiple sources have effect on the target.
+                    // See Issue18282Test for more info.
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/NestedTargetPropertyMappingHolder.java]: nr. 18282.. Did not know we have so many issues :)

File: processor/src/main/java/org/mapstruct/ap/internal/model/NestedTargetPropertyMappingHolder.java [hunk 2]
```diff
@@ -160,6 +160,15 @@ public NestedTargetPropertyMappingHolder build() {
                         groupedByTP.singleTargetReferences.get( targetProperty )
                     );
 
+                    // If there are multi source parameters that are mapped to the target reference or
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/NestedTargetPropertyMappingHolder.java]: Start active phrasing: , so `We need an update method when: 1) condition 1, 2) condition 2).` I can't understand the current sentence

File: processor/src/main/java/org/mapstruct/ap/internal/model/NestedTargetPropertyMappingHolder.java [hunk 3]
```diff
@@ -160,6 +160,15 @@ public NestedTargetPropertyMappingHolder build() {
...
+                    // there are multiple source references to the target reference then we need to force an
+                    // update method. The reason for this is that multiple sources have effect on the target.
+                    // See Issue18282Test for more info.
+                    boolean forceUpdateMethod =
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/NestedTargetPropertyMappingHolder.java]: are there really 2 booleans required to fix this?

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1828/FirstMapper.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.factory.Mappers;
+
+@Mapper
+public interface FirstMapper {
+
+    FirstMapper INSTANCE = Mappers.getMapper( FirstMapper.class );
+
+    @Mapping(target = "name", source = "name")
```
Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1828/FirstMapper.java]: is this required? name based mapping?

File: processor/src/main/java/org/mapstruct/ap/internal/model/NestedTargetPropertyMappingHolder.java (hunk 1 above)
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/model/NestedTargetPropertyMappingHolder.java]: 😂, good catch

File: processor/src/main/java/org/mapstruct/ap/internal/model/NestedTargetPropertyMappingHolder.java (hunk 3 above)
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/model/NestedTargetPropertyMappingHolder.java]: It is one boolean, but I wanted to extract one that forces an update method and the other one with the nested reference

**Comment by filiphr** [comment]
>Are you going to map this as well to the current master?

Already done in PR https://github.com/mapstruct/mapstruct/pull/1931

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1828/FirstMapper.java (hunk 4 above)
Comment (filiphr) [reply processor/src/test/java/org/mapstruct/ap/test/bugs/_1828/FirstMapper.java]: Nope it isn't I copied it from the example

**Comment by filiphr** [comment]
Applied the review comments

**Pull request merged**